community/graphdb-api/src/main/java/org/neo4j/graphdb/IdentifiablePropertyContainer.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphdb;
+
+/**
+ * Provides the ability to obtain the unique id of the implementing
+ * {@link PropertyContainer}.
+ */
+public interface IdentifiablePropertyContainer extends PropertyContainer
```

Comment:
You make some excellent points.


Let’s start with the easiest, being that you are not a fan of the name of this interface. Do you have a suggestion you would prefer?

In terms of the general problem of typically needing to know if the id is for a node or for a relationship in order for the id to be meaningful, that is not only very valid but I imagine particularly much more so valid for the developers working on the actual Neo4j product code base. That said, for developers, such as myself, who are essentially end user consumers of the Neo4j product and are working in unmanaged server extensions or GraphAware modules to create code to create application-level business logic, I find more often than not that I’m able to code to the PropertyContainer interface and do not need to know if the specific implementation is a Node or a Relationship. In fact, by admittedly personal and professional preference, I consider coding to the PropertyContainer interface a good (better?) practice than coding to the implementation, analogous to coding to List rather than coding to ArrayList or Vector (for a trivial example). So while I definitely see the merit, complexities, pros and cons in the design decisions you mention, I think there is valid uses of the getId() being on the PropertyContainer interface for both developers working within Neo4j (such as the core product team) and those working outside it (such as me). I’m undoubtedly heavily influenced by my experience working “downstream” of the core Neo4j developers, where perhaps making use of getId() on the PropertyContainer interface makes more sense for me than it does internally within Neo4j. In sort, perhaps having getId() on the PropertyContainer is more useful to server extension or module developers that it is to developers working on core Neo4j internals.

So, being the contributor of the pull request (PR) to “promote” getId() to the PropertyContainer interface, I unsurprisingly propose that the two topics be split into two separate efforts: that my PR be merged into the core codebase (the short term improvement) and that the conversation of the larger redesign and change in direction take place independently (the long term improvement).

As for the effort of the larger redesign, I am happy to contribute in any way that I can. That said, recognizing that I am a first time contributor of this minor PR and not a member of the core Neo4j product team, I realistically have a fairly narrow scope of possible contribution and would undoubtedly defer to core team members, such as yourself, that have much more familiarity and experience with the code.